### PR TITLE
Add support for AArch32 ARMv8-R MPU

### PR DESCRIPTION
**Pull request by manuargue**
This PR adds support for the AArch32 ARMv8-R MPU and is a continuation of the work done by @julien-massot on [#44078](https://github.com/zephyrproject-rtos/zephyr/pull/44078).

I've kept the original commits, rebased and fixed conflicts (mostly related to includes). On top of that, I've fixed a minor issue in the MPU region check and bumped the timeout for some tests, needed when enabling MPU on `fvp_aemv8r_aarch32`.

All the tests failing with timeout described on [this comment ](https://github.com/zephyrproject-rtos/zephyr/pull/44078#issuecomment-1118101390) are passing. Test enabling fvp (`simulation: armfvp`) in board yml file:
```shell
./scripts/twister -p fvp_baser_aemv8r_aarch32 --runtime-artifact-cleanup --retry-failed 2
Keeping artifacts untouched
INFO    - Using Ninja..
INFO    - Zephyr version: zephyr-v3.1.0-1888-0028e97332953
INFO    - Using 'zephyr' toolchain.
INFO    - Writing JSON report /home/user/zephyrproject/zephyr/twister-out/testplan.json
INFO    - JOBS: 8
INFO    - 1875 test scenarios (1781 configurations) selected, 1468 configurations discarded due to filters.
INFO    - Adding tasks to the queue...
INFO    - Added initial list of jobs to queue
INFO    - Total complete: 1781/1781  100%  skipped: 1468, failed:    0
INFO    - 313 of 1781 test configurations passed (100.00%), 0 failed, 1468 skipped with 0 warnings in 1659.81 seconds
INFO    - In total 20734 test cases were executed, 0 skipped on 0 out of total 473 platforms (0.00%)
INFO    - 258 test configurations executed on platforms, 55 test configurations were only built.
INFO    - Saving reports...
INFO    - Writing JSON report /home/user/zephyrproject/zephyr/twister-out/twister.json
INFO    - Writing xunit report /home/user/zephyrproject/zephyr/twister-out/twister.xml...
INFO    - Writing xunit report /home/user/zephyrproject/zephyr/twister-out/twister_report.xml...
INFO    - Run completed
```

### CODE REVIEW

File: arch/arm/core/aarch32/mpu/arm_mpu_v8_internal.h
Comment (andyross) [inline arch/arm/core/aarch32/mpu/arm_mpu_v8_internal.h:42]:
```
__ASSERT(function_names_should_provide_meaningful_abstraction())
maybe_just_use_the_struct_directly()
save_30_lines_too();
```

File: arch/arm/core/aarch32/mpu/arm_mpu_v8_internal.h
Comment (andyross) [inline arch/arm/core/aarch32/mpu/arm_mpu_v8_internal.h:83]: Doesn't DSB already subsume an instruction barrier?  Seems weird to actually need both.

File: tests/posix/common/testcase.yaml
Comment (andyross) [inline tests/posix/common/testcase.yaml:5]: Ouch.  Is 10 minutes really needed?

**Review by andyross — APPROVED** [review]
Not super expert on ARM nor the mmu layer, but no red flags I can see, just nitpicks.

File: tests/posix/common/testcase.yaml
Comment (manuargue) [reply tests/posix/common/testcase.yaml:5]: On my modest setup, sometimes it fails with timeout if less time is used. An alternative would be to reduce the [BOUNCES count](https://github.com/zephyrproject-rtos/zephyr/blob/main/tests/posix/common/src/pthread.c#L19) for this platform to an acceptable value so the test still makes sense. Although an increase in the timeout would be still needed I think.

File: arch/arm/core/aarch32/mpu/arm_mpu_v8_internal.h
Comment (manuargue) [reply arch/arm/core/aarch32/mpu/arm_mpu_v8_internal.h:42]: > maybe_just_use_the_struct_directly()
save_30_lines_too();

The idea of this MPU internal abstraction is to isolate the specifics for Armv8-M and ARMv8-R AArch32 in order to reuse the rest of the common code. For ARMv8-R AArch32, the MPU registers must be accessed using the `mrc/mcr` hence is not possible to use the same implementation as for Armv8-M.

File: arch/arm/core/aarch32/mpu/arm_mpu_v8_internal.h
Comment (manuargue) [reply arch/arm/core/aarch32/mpu/arm_mpu_v8_internal.h:83]: AFAIK, `DSB` has no effect on pre-fetching of instructions and `ISB` will flush the pipelines to fetch next instructions from cache or memory.

**Review by julien-massot — APPROVED** [review]
Thanks @manuargue for handling this PR.

File: tests/posix/common/testcase.yaml
Comment (povergoing) [reply tests/posix/common/testcase.yaml:5]: I encountered a similar issue, you may ref https://github.com/zephyrproject-rtos/zephyr/pull/40575 to add a platform-specific test timeout, thus your change will not affect other platforms.

File: tests/kernel/workq/critical/testcase.yaml
Comment (povergoing) [inline tests/kernel/workq/critical/testcase.yaml:3]: is the default timeout 60?

Comment (carlescufi) [reply tests/kernel/workq/critical/testcase.yaml:3]: @povergoing in order not to block this PR, would you mind approving if this is your only concern? Then @manuargue can submit a follow-up PR to remove this line.

Comment (povergoing) [reply tests/kernel/workq/critical/testcase.yaml:3]: sure, No problem for me

**Review by povergoing — APPROVED** [review]

File: tests/kernel/workq/critical/testcase.yaml
Comment (manuargue) [reply tests/kernel/workq/critical/testcase.yaml:3]: ok, I will submit a follow-up PR for implementing suggestion on this comment https://github.com/zephyrproject-rtos/zephyr/pull/47579#discussion_r917580682

**Pull request merged**
